Add `compat` configuration options reference (#4352)
tarantool/doc: doc/reference/configuration/configuration_reference.rst

* :ref:`compat.box_info_cluster_meaning <configuration_reference_compat_yaml_pretty>`
* :ref:`compat.box_session_push_deprecation <configuration_reference_compat_session_push>`
* :ref:`compat.box_space_execute_priv <configuration_reference_compat_space_execute>`
* :ref:`compat.box_space_max <configuration_reference_compat_space_max>`
* :ref:`compat.box_tuple_extension <configuration_reference_compat_tuple_extension>`
* :ref:`compat.box_tuple_new_vararg <configuration_reference_compat_tuple_new>`
* :ref:`compat.c_func_iproto_multireturn <configuration_reference_compat_iproto_multireturn>`
* :ref:`compat.fiber_channel_close_mode <configuration_reference_compat_fiber_channel>`
* :ref:`compat.fiber_slice_default <configuration_reference_compat_cluster_meaning>`
* :ref:`compat.json_escape_forward_slash <configuration_reference_compat_json_escape>`
* :ref:`compat.sql_priv <configuration_reference_compat_sql_priv>`
* :ref:`compat.sql_seq_scan_default <configuration_reference_compat_sql_scan>`
* :ref:`compat.yaml_pretty_multiline <configuration_reference_compat_yaml_pretty>`

.. _configuration_reference_compat_binary_decoding:

.. confval:: compat.binary_data_decoding

>     Defines how to store binary data fields in Lua after decoding:

Reviewer: Let's use [the imperative mood](https://www.tarantool.io/en/doc/latest/contributing/docs/api/#language-of-the-general-description) as in other config/API descriptions:

```suggestion
    Define how to store binary data fields in Lua after decoding:
```

    -   ``new``: as varbinary objects
    -   ``old``: as plain strings

    |
    | Type: string
    | Possible values: 'new', 'old'
    | Default: 'new'
    | Environment variable: TT_COMPAT_BINARY_DATA_DECODING

.. _configuration_reference_compat_replication_timeout:

.. confval:: compat.box_cfg_replication_sync_timeout

>     Sets a default replication sync timeout: 0 (`new`) or 300 seconds (`old`)

Reviewer: ```suggestion
    Set a default replication sync timeout: 0 (`new`) or 300 seconds (`old`).
```

    .. important::

        This value is set during the initial ``box.cfg{}`` call and cannot be changed later.

    See also: :ref:`compat-option-replication-timeout`

    |
    | Type: string
    | Possible values: 'new', 'old'
    | Default: 'new'
    | Environment variable: TT_COMPAT_BOX_CFG_REPLICATION_SYNC_TIMEOUT

.. _configuration_reference_compat_error_serialize:

> .. confval:: compat.box_error_serialize_verbose

Reviewer: It looks like this option first appeared in [3.1.0](https://github.com/tarantool/tarantool/releases/tag/3.1.0). Let's add `Since` here.

>     Controls the verbosity of :ref:`error objects <box_error-error_object>` serialization:

Reviewer: ```suggestion
    Set the verbosity of :ref:`error objects <box_error-error_object>` serialization:
```

    -   ``new``: serialize the error message together with other potentially useful fields
    -   ``old``: serialize only the error message

    |
    | Type: string
    | Possible values: 'new', 'old'
    | Default: 'old'
    | Environment variable: TT_COMPAT_BOX_ERROR_SERIALIZE_VERBOSE

.. _configuration_reference_compat_error_unpack:

> .. confval:: compat.box_error_unpack_type_and_code

Reviewer: Since [3.1.0](https://github.com/tarantool/tarantool/releases/tag/3.1.0).

    Whether to show error fields in ``box.error.unpack()``:

    -   ``new``: do not show ``base_type`` and ``custom_type`` fields; do not show

>         the ``code`` field if it is 0. Note that ``base_type`` is still accessible for error object.

Reviewer: ```suggestion
        the ``code`` field if it is 0. Note that ``base_type`` is still accessible for an error object.
```

    Whether ``box.info.cluster`` should show the current replica set (`old`) or
    the whole cluster with all its replica sets (`new`).

    |
    | Type: string
    | Possible values: 'new', 'old'
    | Default: 'new'
    | Environment variable: TT_COMPAT_BOX_INFO_CLUSTER_MEANING

.. _configuration_reference_compat_session_push:

.. confval:: compat.box_session_push_deprecation

    Whether to raise errors on attempts to call the deprecated function ``box.session.push``:

    -   ``new``: raise an error
    -   ``old``: do not raise an error

    |
    | Type: string
    | Possible values: 'new', 'old'
    | Default: 'old'
    | Environment variable: TT_COMPAT_BOX_SESSION_PUSH_DEPRECATION

.. _configuration_reference_compat_space_execute:

.. confval:: compat.box_space_execute_priv

    Whether the ``execute`` privilege can be granted on spaces:

    -   ``new``: an error is raised
    -   ``old``: the privilege can be granted with no actual effect

    |
    | Type: string
    | Possible values: 'new', 'old'
    | Default: 'new'
    | Environment variable: TT_COMPAT_BOX_SPACE_EXECUTE_PRIV

.. _configuration_reference_compat_space_max:

.. confval:: compat.box_space_max

>     Controls the max space id (``box.schema.SPACE_MAX``):

Reviewer: ```suggestion
    Set the maximum space identifier (``box.schema.SPACE_MAX``):
```

Let's also update a corresponding value in the `Limitations` topic: https://www.tarantool.io/en/doc/latest/book/box/limitations/

    -   ``old``: tuples with formats are encoded as ``MP_ARRAY``

    |
    | Type: string
    | Possible values: 'new', 'old'
    | Default: 'new'
    | Environment variable: TT_COMPAT_BOX_TUPLE_EXTENSION

.. _configuration_reference_compat_tuple_new:

.. confval:: compat.box_tuple_new_vararg

    Controls how ``box.tuple.new`` interprets an argument list:

    -   ``new``: as a value with a tuple format
    -   ``old``: as an array of tuple fields

    |
    | Type: string
    | Possible values: 'new', 'old'
    | Default: 'new'
    | Environment variable: TT_COMPAT_BOX_TUPLE_NEW_VARARG

.. _configuration_reference_compat_iproto_multireturn:

.. confval:: compat.c_func_iproto_multireturn

    Controls wrapping of multiple results of a stored C function when returning them via iproto:

    -   ``new``: return without wrapping (consistently with a local call via ``box.func``)
    -   ``old``: wrap results into a MessagePack array

    |
    | Type: string
    | Possible values: 'new', 'old'
    | Default: 'new'
    | Environment variable: TT_COMPAT_C_FUNC_IPROTO_MULTIRETURN

.. _configuration_reference_compat_fiber_channel:

.. confval:: compat.fiber_channel_close_mode

>     Defines the behavior of filber channels after closing:

Reviewer: ```suggestion
    Define the behavior of fiber channels after closing:
```
